### DISCUSSION TITLE
Edit variables in a matched pattern

### DISCUSSION THREAD

**Discussion by goodov** [ast-grep/ast-grep#1527] [General]
Is there a way to match part of code with meta-variables and let `ast-grep` rewrite the variables while keeping other parts of the matched code untouched? The use case for this is syntax-aware patching where I want to match the exact AST surrounding the code I want to patch.

The issue I'm facing right now is that I should rewrite the template as is to add something new in the `fix` part, instead of just modifying the variable.

Consider this example ([playground link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImphdmFzY3JpcHQiLCJxdWVyeSI6ImlmICh0cnVlKSB7XHJcbiAgICAkJCRcclxufSBlbHNlIHtcclxuICAgICQkJFxyXG59IiwicmV3cml0ZSI6IiIsInN0cmljdG5lc3MiOiJzaWduYXR1cmUiLCJzZWxlY3RvciI6IiIsImNvbmZpZyI6InJ1bGU6XG4gIHBhdHRlcm46XG4gICAgY29udGV4dDogfC1cbiAgICAgIGlmICh0cnVlKSB7XG4gICAgICAgICQkJFxuICAgICAgICBsZXQgYSA9IGdldEEoKTtcbiAgICAgICAgJCQkSU5TRVJUSU9OX1BPSU5UXG4gICAgICAgIGlmIChhID09IDMzMykgeyAkJCQgfVxuICAgICAgICAkJCRcbiAgICAgIH1cbiAgICBzdHJpY3RuZXNzOiBjc3RcbmZpeDogfC1cbiAgaWYgKHRydWUpIHtcbiAgICBjb25zdCBub3RUaGlzID0gJ2NvbnNvbGUubG9nKFwibm90IG1lXCIpJ1xuICAgIGxldCBhID0gZ2V0QSgpO1xuICAgIGlmIChhID09IDExMSkge1xuICAgICAgcmV0dXJuIGZhbHNlO1xuICAgIH1cbiAgICBpZiAoYSA9PSAzMzMpIHtcbiAgICAgIHJldHVybiBmYWxzZTtcbiAgICB9XG4gICAgaWYgKGEgPT0gNDQ0KSB7XG4gICAgICByZXR1cm4gdHJ1ZTtcbiAgICB9XG4gIH1cbiIsInNvdXJjZSI6Ii8vIGNvbnNvbGUubG9nKCkgd2lsbCBiZSBtYXRjaGVkIGJ5IHBhdHRlcm4hXG4vLyBjbGljayBkaWZmIHRhYiB0byBzZWUgcmV3cml0ZS5cblxuaWYgKHRydWUpIHtcbiAgY29uc3Qgbm90VGhpcyA9ICdjb25zb2xlLmxvZyhcIm5vdCBtZVwiKSdcbiAgbGV0IGEgPSBnZXRBKCk7XG4gIGlmIChhID09IDMzMykge1xuICAgIHJldHVybiBmYWxzZTtcbiAgfVxuICBpZiAoYSA9PSA0NDQpIHtcbiAgICByZXR1cm4gdHJ1ZTtcbiAgfVxufSJ9)):
```javascript
if (true) {
  const notThis = 'console.log("not me")'
  const a = getA();
  if (a == 333) {
    return false;
  }
  if (a == 444) {
    return true;
  }
}
```

I want to add new code after `const a = getA();` line, but before `if (a == 333) {`. Right now to do this I had to write this config:
```yaml
rule:
  pattern:
    context: |-
      if (true) {
        $$$
        const a = getA();
        $$$INSERTION_POINT
        if (a == 333) { $$$ }
        $$$
      }
    strictness: cst
fix: |-
  if (true) {
    const notThis = 'console.log("not me")'
    const a = getA();
    if (a == 111) {  // This is the new condition added in the "fix" block.
      return false;
    }
    if (a == 333) {
      return false;
    }
    if (a == 444) {
      return true;
    }
  }
```

As you can see, I had to mostly duplicate the matched code in the `fix` part. What I'd like to have is something like `edit` functionality:
```yaml
edit:
  - replace: $$$INSERTION_POINT
    with: |-
      if (a == 111) {
        return false;
      }
      $$$INSERTION_POINT
```

Does `ast-grep` support this in some way? What are my options here? Any advice would be greatly appreciated!

**Comment by HerringtonDarkholme** [comment]
You can use rewriter and replace
https://ast-grep.github.io/guide/rewrite/rewriter.html
